core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala

```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
-                case Some(metadata) =>
-                  // do not synchronize on this metadata since it will only be accessed by the sender thread
-                  metadata.topicPartitions -= topicPartition
+          case Some(epochAndMetadata) =>
+            val retryPartitions: mutable.Set[TopicPartition] = mutable.Set.empty[TopicPartition]
+            val errors = writeTxnMarkerResponse.errors(txnMarker.producerId)
+
+            if (errors == null)
+              throw new IllegalStateException(s"WriteTxnMarkerResponse does not contain expected error map for pid ${txnMarker.producerId}")
+
+            epochAndMetadata synchronized {
+              for ((topicPartition: TopicPartition, error: Errors) <- errors) {
+                error match {
+                  case Errors.NONE =>
+                    epochAndMetadata.transactionMetadata.topicPartitions -= topicPartition
+
+                  case Errors.UNKNOWN_TOPIC_OR_PARTITION | Errors.NOT_LEADER_FOR_PARTITION |
+                       Errors.NOT_ENOUGH_REPLICAS | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND =>
+                    retryPartitions += topicPartition
+
+                  case other =>
+                    throw new IllegalStateException(s"Unexpected error ${other.exceptionName} while sending txn marker for $transactionalId")
+                }
               }
-            case Errors.UNKNOWN_TOPIC_OR_PARTITION | Errors.NOT_LEADER_FOR_PARTITION |
-                 Errors.NOT_ENOUGH_REPLICAS | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND =>
-              retryPartitions += topicPartition
-            case _ =>
-              throw new IllegalStateException("Writing txn marker request failed permanently for pid " + txnMarker.producerId())
-          }
+            }
 
-          if (retryPartitions.nonEmpty) {
-            // re-enqueue with possible new leaders of the partitions
-            transactionMarkerChannel.addRequestToSend(
-              txnTopicPartition,
-              txnMarker.producerId(),
-              txnMarker.producerEpoch(),
-              txnMarker.transactionResult,
-              txnMarker.coordinatorEpoch(),
-              retryPartitions.toSet)
-          }
-          val completed = txnMarkerPurgatory.checkAndComplete(txnMarker.producerId())
-          trace(s"Competed $completed transactions for producerId ${txnMarker.producerId()}")
+            if (retryPartitions.nonEmpty) {
+              // re-enqueue with possible new leaders of the partitions
+              txnMarkerChannel.addTxnMarkersToSend(
+                transactionalId,
+                txnMarker.producerId(),
+                txnMarker.producerEpoch(),
+                txnMarker.transactionResult,
+                txnMarker.coordinatorEpoch(),
+                retryPartitions.toSet)
+            } else {
+              txnMarkerPurgatory.checkAndComplete(txnMarker.producerId)
```

Comment:
Hmm, is this right? We watch on transactionalId, not producerId. Also, since txnMarkerPurgatory is passed around to different classes, it's bit hard to track who is calling checkAndComplete on this purgatory. An alternative way is to expose an access method to txnMarkerPurgatory in TransactionChannelManager and let all classes call that method.